numba/cuda/cudaimpl.py

```diff
@@ -7,135 +7,117 @@
-    return res
 
+@lower_attr(types.Module(cuda), 'gridDim')
+def cuda_gridDim(context, builder, sig, args):
+    return initialize_dim3(builder, 'nctaid')
 
-@lower('ptx.gridsize.2d', types.intp)
-def ptx_gridsize2d(context, builder, sig, args):
-    assert len(args) == 1
-    ntidx = nvvmutils.call_sreg(builder, "ntid.x")
-    nctaidx = nvvmutils.call_sreg(builder, "nctaid.x")
 
-    ntidy = nvvmutils.call_sreg(builder, "ntid.y")
-    nctaidy = nvvmutils.call_sreg(builder, "nctaid.y")
+@lower_attr(types.Module(cuda), 'laneid')
+def cuda_laneid(context, builder, sig, args):
+    return nvvmutils.call_sreg(builder, 'laneid')
 
-    r1 = builder.mul(ntidx, nctaidx)
-    r2 = builder.mul(ntidy, nctaidy)
-    return cgutils.pack_array(builder, [r1, r2])
 
+@lower_attr(types.Module(cuda), 'warpsize')
+def cuda_warpsize(context, builder, sig, args):
+    return nvvmutils.call_sreg(builder, 'warpsize')
 
-@lower('ptx.gridsize.3d', types.intp)
-def ptx_gridsize3d(context, builder, sig, args):
-    assert len(args) == 1
-    ntidx = nvvmutils.call_sreg(builder, "ntid.x")
-    nctaidx = nvvmutils.call_sreg(builder, "nctaid.x")
 
-    ntidy = nvvmutils.call_sreg(builder, "ntid.y")
-    nctaidy = nvvmutils.call_sreg(builder, "nctaid.y")
+@lower_attr(dim3, 'x')
+def dim3_x(context, builder, sig, args):
+    return builder.extract_value(args, 0)
 
-    ntidz = nvvmutils.call_sreg(builder, "ntid.z")
-    nctaidz = nvvmutils.call_sreg(builder, "nctaid.z")
 
-    r1 = builder.mul(ntidx, nctaidx)
-    r2 = builder.mul(ntidy, nctaidy)
-    r3 = builder.mul(ntidz, nctaidz)
-    return cgutils.pack_array(builder, [r1, r2, r3])
+@lower_attr(dim3, 'y')
+def dim3_y(context, builder, sig, args):
+    return builder.extract_value(args, 1)
 
 
-# -----------------------------------------------------------------------------
+@lower_attr(dim3, 'z')
+def dim3_z(context, builder, sig, args):
+    return builder.extract_value(args, 2)
 
-def ptx_sreg_template(sreg):
-    def ptx_sreg_impl(context, builder, sig, args):
-        assert not args
-        return nvvmutils.call_sreg(builder, sreg)
 
-    return ptx_sreg_impl
+@lower(cuda.grid, types.int32)
+def cuda_grid(context, builder, sig, args):
+    restype = sig.return_type
+    if restype == types.int32:
+        return nvvmutils.get_global_id(builder, dim=1)
+    elif isinstance(restype, types.UniTuple):
+        ids = nvvmutils.get_global_id(builder, dim=restype.count)
+        return cgutils.pack_array(builder, ids)
+    else:
+        raise ValueError('Unexpected return type %s from cuda.grid' % restype)
 
 
-# Dynamic create all special register
-for sreg in nvvmutils.SREG_MAPPING.keys():
-    lower(sreg)(ptx_sreg_template(sreg))
+@lower(cuda.gridsize, types.int32)
+def cuda_gridsize(context, builder, sig, args):
+    restype = sig.return_type
 
+    ntidx = nvvmutils.call_sreg(builder, "ntid.x")
+    nctaidx = nvvmutils.call_sreg(builder, "nctaid.x")
+    nx = builder.mul(ntidx, nctaidx)
```

Comment:
can probably make this into a little function and call for x, y, z? not essential

Reply:
done. That function will probably be handy for the implementation of future functions, so might as well have it now.